exporter/otlpexporter/arrow.go

```diff
@@ -0,0 +1,469 @@
+			running--
+
+			// None of the streams were able to connect to
+			// an Arrow endpoint.
+			if running == 0 {
+				ae.ready.downgrade()
+			}
+
+		case <-bgctx.Done():
+			// We are shutting down.
+			return
+		}
+	}
+}
+
+// runArrowStream begins one gRPC stream using a child of the background context.
+// If the stream connection is successful, this goroutine starts another goroutine
+// to call writeStream() and performs readStream() itself.  When the stream shuts
+// down this call synchronously waits for and unblocks the consumers.
+func (ae *arrowExporter) runArrowStream(bgctx context.Context) {
+	ctx, cancel := context.WithCancel(bgctx)
+
+	stream := &arrowStream{
+		toWrite:  make(chan writeItem, 1),
+		producer: batchEvent.NewProducer(),
+		waiters:  map[string]chan error{},
+		cancel:   cancel,
+	}
+
+	defer func() {
+		ae.wg.Done()
+		cancel()
+		ae.returning <- stream
+	}()
+
+	sc, err := ae.client.EventStream(ctx, ae.exporter.callOptions...)
+	if err != nil {
+		// TODO: only when this is a permanent (e.g., "no such
+		// method") error, downgrade to standard OTLP.
+		// Returning with stream.client == nil signals the
+		// lack of an Arrow stream endpoint.  When all the
+		// streams return with .client == nil, the ready
+		// channel will be closed.
+		ae.exporter.settings.Logger.Error("cannnot start event stream", zap.Error(err))
+		return
+	}
+	// Setting .client != nil indicates that the endpoint was valid,
+	// streaming may start.  When this stream finishes, it will be
+	// restarted.
+	stream.client = sc
+
+	var ww sync.WaitGroup
+
+	ww.Add(1)
+	ae.wg.Add(1)
+	go ae.writeStream(ctx, stream, &ww)
+
+	if err := ae.readStream(ctx, stream); err != nil {
+		// TODO: should this log even an io.EOF error?
+		ae.exporter.settings.Logger.Error("arrow recv", zap.Error(err))
+	}
+
+	// Wait for the writer to ensure that all waiters are known.
+	ww.Wait()
+
+	// The reader and writer have both finished; respond to any
+	// outstanding waiters.
+	for _, ch := range stream.waiters {
+		// Note: exporterhelper will retry.
+		// TODO: Would it be better to handle retry in this directly?
+		ch <- status.Error(codes.Aborted, "stream is restarting")
+	}
+}
+
+// getArrowStream returns the first-available stream.  This returns nil
+// when the consumer should fall back to standard OTLP.
+func (e *exporter) getArrowStream(ctx context.Context) (*arrowStream, error) {
+	if e.arrow == nil {
+		return nil, nil
+	}
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case stream := <-e.arrow.ready.readyChannel():
+		return stream, nil
+	}
+}
+
+// shutdown returns when all Arrow-associated goroutines have returned.
+func (ae *arrowExporter) shutdown(ctx context.Context) error {
+	ae.cancel()
+	ae.wg.Wait()
+	return nil
+}
+
+// setBatchChannel places a waiting consumer's batchID into the waiters map, where
+// the stream reader may find it.
+func (stream *arrowStream) setBatchChannel(batchID string, errCh chan error) {
```

Comment:
nit: can you group the `arrowStream` and `arrowExporter` methods?

Reply:
Done -- this exploded the size of the next PR.